gui

```diff
@@ -191,570 +438,258 @@ runonce <<"EOF"
+#In Xlunch mode, the window must be centeredm but xlunch only allows absolute offsets.
+{
+  #determine screen_width and screen_height
+  xrandr_out="$(xrandr | grep "HDMI-1" | awk '{print $4}' | tr 'x+' ' ')"
+  screen_width="$(awk '{print $1}' <<<"$xrandr_out")"
+  screen_height="$(awk '{print $2}' <<<"$xrandr_out")"
+  unset xrandr_out
   
-  #terminal title text
-  linecount="$(echo "$app" | wc -l)"
-  if [ $linecount -eq 1 ];then
-    title="Installing $app"
-  elif [ $linecount -lt 4 ];then
-    title="Installing $(echo -n "$app" | tr '\n' '|' | sed 's/|/, /g')"
+  if [[ "$guimode" == yad* ]];then
+    #total dimensions for both yad windows side by side
+    height=400
+    width=600
+    #how far down from the top of screen should both windows be?
+    yoffset=$(((screen_height/2)-(height/2)))
+    
+    #width of first window
+    width1=250
+    #width of second window
+    width2=$((width - width1))
+    
+    #screen offsets for window 1
+    xoffset1=$(((screen_width/2)-(width/2)))
+    #screen offsets for window 2
+    xoffset2=$(((screen_width/2)-(width/2)+width1))
+    
+    #set completed location arguments for both yad windows
+    geometry1="--geometry=${width1}x${height}+${xoffset1}+${yoffset}"
+    geometry2="--geometry=${width2}x${height}+${xoffset2}+${yoffset}"
+    
+  elif [[ "$guimode" == xlunch* ]];then
+    #desired dimensions for xlunch window
+    height=700
+    width=800
+    #determine 
+    xoffset=$(((screen_width/2)-(width/2)))
+    yoffset=$(((screen_height/2)-(height/2)))
   else
-    title="Installing $linecount apps"
+    error "Unrecognized app list style '$guimode'!"
   fi
-  "${DIRECTORY}/etc/terminal-run" '
-    DIRECTORY="'"$DIRECTORY"'"
-    source "${DIRECTORY}/api"
-    generate_logo
-    if "${DIRECTORY}/manage" multi-install "'"$app"'" ;then
-      echo -e "\nClosing in 30 seconds."
-      sleep 30
-    else
-      echo -e "\nClose this window to exit."
-      read enter #technically you could press Enter to exit.
-    fi
-    ' "$title"
-  
-  #re-preload all categories in background
-  "${DIRECTORY}/etc/preload-daemon" "$guimode" &>/dev/null &
 }
 
-uninstall() {
-  app="$1" #one app name per line
+#Compile xlunch if required
+if [[ "$guimode" == xlunch* ]] && ([ ! -d "${DIRECTORY}/xlunch" ] || [ ! -f /usr/bin/xlunch ]);then
+  #signal files
+  rm -f /tmp/xlunchfailed /tmp/xlunchfinished /tmp/terminalexit
+  echo '' > /tmp/terminalexit
+  "${DIRECTORY}/etc/terminal-run" "
+    function error {
+      echo -e "\""\e[91m$1\e[39m"\""
+      echo 'Close this terminal to exit.'
+      echo '' > /tmp/xlunchfailed
+      sleep infinity
+    }
+    trap 'echo "\"""\"" > /tmp/terminalexit' EXIT
+    
+    #uninstall xlunch first
+    sudo rm -rf /etc/xlunch /usr/share/xlunch /usr/bin/xlunch /usr/bin/genentries /usr/bin/updateentries /usr/bin/genentries.desktop.sh /usr/share/icons/hicolor/48x48/apps/xlunch_ghost.png /usr/share/icons/hicolor/48x48/apps/xlunch.png /usr/share/applications/genentries.desktop
```

Comment:
same issue as I pointed out before. most users don't have passwordless sudo (its super insecure and should never be used).
this sudo is used in the gui in a hidden terminal prompt if used from the pi-apps desktop icon, so this script will always fail to run.

Reply:
This is actually running inside of `terminal-run`, so if a password is required then the user will be prompted for it.